### PR TITLE
ENH: Add `float` wrapping for itkSymmetricSecondRankTensor

### PR DESCRIPTION
**Pull request by tbirdso**
Adds `float` wrapping for `itk::SymmetricSecondRankTensor` as well as its usage in `itk::Image` and `itk::ImageToImageFilter`.

Motivated by updates in [ITKStrain](https://github.com/KitwareMedical/ITKStrain/pull/42).

## PR Checklist
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

File: Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap [hunk 1]
```diff
@@ -32,13 +32,16 @@ itk_wrap_class("itk::ImageToImageFilter" POINTER)
...
-      itk_wrap_template("${ITKM_ISSRT${ITKM_D}${d}${d}}${ITKM_I${t}${d}}" "${ITKT_ISSRT${ITKM_D}${d}${d}}, ${ITKT_I${t}${d}}")
+    foreach(t2 ${WRAP_ITK_REAL})
+      foreach(t ${WRAP_ITK_SCALAR})
+        # scalar <-> SymmetricSecondRankTensor
+        itk_wrap_template("${ITKM_I${t}${d}}${ITKM_ISSRT${ITKM_${t2}}${d}${d}}" "${ITKT_I${t}${d}}, ${ITKT_ISSRT${ITKM_${t2}}${d}${d}}")
+        itk_wrap_template("${ITKM_ISSRT${ITKM_${t2}}${d}${d}}${ITKM_I${t}${d}}" "${ITKT_ISSRT${ITKM_${t2}}${d}${d}}, ${ITKT_I${t}${d}}")
+      endforeach()
+      # SymmetricSecondRankTensor -> SymmetricSecondRankTensor
+      itk_wrap_template("${ITKM_ISSRT${ITKM_${t2}}${d}${d}}${ITKM_ISSRT${ITKM_${t2}}${d}${d}}"
+                        "${ITKT_ISSRT${ITKM_${t2}}${d}${d}}, ${ITKT_ISSRT${ITKM_${t2}}${d}${d}}")
```
Comment (Leengit) [inline Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap:43]: We can also support the likes of SymmetricSecondRankTensor[float] -> SymmetricSecondRankTensor[double] by instead using:
```cmake
      # SymmetricSecondRankTensor -> SymmetricSecondRankTensor
      foreach(t3 ${WRAP_ITK_REAL})
        itk_wrap_template("${ITKM_ISSRT${ITKM_${t3}}${d}${d}}${ITKM_ISSRT${ITKM_${t2}}${d}${d}}"
                          "${ITKT_ISSRT${ITKM_${t3}}${d}${d}}, ${ITKT_ISSRT${ITKM_${t2}}${d}${d}}")
      endforeach()
```

Comment (tbirdso) [reply Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap]: Thanks @Leengit , this is a good idea. I've added the changes, will squash commits once CI completes.

**Comment by tbirdso** [comment]
Following up on build failures here [https://open.cdash.org/viewBuildError.php?type=1&buildid=7287021](https://open.cdash.org/viewBuildError.php?type=1&buildid=7287021), `itkSymmetricEigenAnalysisImageFilter` is failing to find updated wrappings:

> itkSymmetricEigenAnalysisImageFilter: warning(4): ITK type not wrapped, or currently not known: itk::ImageToImageFilter< itk::Image< itk::SymmetricSecondRankTensor< double, 2 >, 2 >, itk::Image< itk::SymmetricSecondRankTensor< double, 2 >, 2 >

I'm having trouble reproducing this on Windows, not getting a similar build warning in VS and it looks like wrappings are being generated correctly:

>\> itk.ImageToImageFilter.GetTypes()
<itkTemplate itk::ImageToImageFilter\>
Options:
...
[<class 'itk.itkImagePython.itkImageSSRTD22'>, <class 'itk.itkImagePython.itkImageSSRTD22'>]
...

>\> itk.SymmetricEigenAnalysisImageFilter.GetTypes()
<itkTemplate itk::SymmetricEigenAnalysisImageFilter\>
Options:
  [<class 'itk.itkImagePython.itkImageSSRTD22'\>,]
  [<class 'itk.itkImagePython.itkImageSSRTD33'\>,]

Relevant changes [here](https://github.com/InsightSoftwareConsortium/ITK/pull/2593/files#diff-c776cb5c228a3c856e7e3423476e2e5a3078437603166f2d14288ddd6e131637R41).

**Comment by dzenanz** [comment]
There are issues:
```text
itkSymmetricEigenAnalysisImageFilter: warning(4): ITK type not wrapped, or currently not known: itk::ImageToImageFilter< itk::Image< itk::SymmetricSecondRankTensor< double, 2 >, 2 >, itk::Image< itk::SymmetricSecondRankTensor< double, 2 >, 2 > >
itkSymmetricEigenAnalysisImageFilter: warning(4): ITK type not wrapped, or currently not known: itk::ImageToImageFilter< itk::Image< itk::SymmetricSecondRankTensor< double, 3 >, 3 >, itk::Image< itk::SymmetricSecondRankTensor< double, 3 >, 3 > >
Wrapping/Typedefs/itkSymmetricEigenAnalysisImageFilter.i:100: Warning 401: Nothing known about base class 'itk::ImageToImageFilter< itk::Image< itk::SymmetricSecondRankTensor< double,2 >,2 >,itk::Image< itk::SymmetricSecondRankTensor< double,2 >,2 > >'. Ignored.
```
etc.

**Comment by dzenanz** [comment]
I took a look, and I don't know what is wrong. Perhaps take another look at https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/ImageIntensity/wrapping/itkSymmetricEigenAnalysisImageFilter.wrap?

**Comment by tbirdso** [comment]
Thanks for taking a look @dzenanz. I've pushed a WIP commit to see if we get warnings about wrapping duplication if we try to wrap `ImageToImageFilter` again just before wrapping `SymmetricEigenAnalysisImageFilter`.

I was wondering if there was any way for this to be a race condition between the two image filter wrappings, but in the [failed build logs](https://dev.azure.com/itkrobotlinuxpython/59377e31-8b02-46ea-9d49-32b6bb4b2da6/_apis/build/builds/4756/logs/11) `ImageToImageFilterB.wrap` is built before `SymmetricEigenAnalysisImageFilter` so that doesn't appear to be the case?

**Comment by tbirdso** [comment]
Still working on this one, haven't been able to reproduce locally on either Windows or Ubuntu platforms using default build settings. Maybe there is something different in CI settings?

File: Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap [hunk 2]
```diff
@@ -32,13 +32,18 @@ itk_wrap_class("itk::ImageToImageFilter" POINTER)
   itk_wrap_image_filter_combinations("${WRAP_ITK_COMPLEX_REAL}" "${WRAP_ITK_SCALAR}")
 
   foreach(d ${ITK_WRAP_IMAGE_DIMS})
-    foreach(t ${WRAP_ITK_SCALAR})
-      # scalar <-> SymmetricSecondRankTensor
-      itk_wrap_template("${ITKM_I${t}${d}}${ITKM_ISSRT${ITKM_D}${d}${d}}" "${ITKT_I${t}${d}}, ${ITKT_ISSRT${ITKM_D}${d}${d}}")
```
Comment (thewtex) [inline Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap:37]: @tbirdso we were always wrapping for `double` before.

File: Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap [hunk 3]
```diff
@@ -32,13 +32,18 @@ itk_wrap_class("itk::ImageToImageFilter" POINTER)
...
-      itk_wrap_template("${ITKM_ISSRT${ITKM_D}${d}${d}}${ITKM_I${t}${d}}" "${ITKT_ISSRT${ITKM_D}${d}${d}}, ${ITKT_I${t}${d}}")
+    foreach(t2 ${WRAP_ITK_REAL})
```
Comment (thewtex) [inline Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap:35]: We still need to always wrap for `double`, so, e.g.:

```
UNIQUE(real_types "${WRAP_ITK_REAL};D")

[...]
foreach(t2 ${real_types})
```

File: Modules/Core/Common/wrapping/itkImageSource.wrap [hunk 4]
```diff
@@ -8,7 +8,9 @@ itk_wrap_class("itk::ImageSource" POINTER)
     foreach(type ${to_types})
       itk_wrap_template("${ITKM_VI${type}${d}}"  "${ITKT_VI${type}${d}}")
     endforeach()
-    itk_wrap_template("${ITKM_ISSRT${ITKM_D}${d}${d}}"  "${ITKT_ISSRT${ITKM_D}${d}${d}}")
+    foreach(t ${WRAP_ITK_REAL})
```
Comment (thewtex) [inline Modules/Core/Common/wrapping/itkImageSource.wrap:11]: Always wrap for `double` as below.

**Review by thewtex — CHANGES_REQUESTED** [review]

File: Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap (hunk 2 above)
Comment (tbirdso) [reply Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap:37]: Thanks for taking a look @thewtex, it didn't occur to me that we're not wrapping for `double` by default.

File: Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap (hunk 3 above)
Comment (tbirdso) [reply Modules/Core/Common/wrapping/itkImageToImageFilterB.wrap]: Fixed!

**Comment by tbirdso** [comment]
Rebased on `master`.

**Comment by tbirdso** [comment]
> I think you will want to define real_types in some (all?) of the other files too, for use instead of WRAP_ITK_REAL.

Thank you @Leengit, fixed now in `ImageSource.wrap` and `GenerateImageSource.wrap`. Continuing to manually check for `float` wrapping in `WrapITKTypes.cmake` to stay consistent with styling in that file.

**Review by thewtex — APPROVED** [review]
💚

**Pull request merged**